Fix EVP_Digest{Sign,Verify}Final() and EVP_Digest{Sign,Verify}() for provider only keys

levitte opened the pull request:
For provider only keys where the initialization didn't catch, we may
end up crashing because the legacy code path didn't check that it had
support carefully enough.  This only happens if the caller didn't
check if initialization worked or not.

For the one-shot case, it's very simply handling the case where the
key has no legacy implementation an fall back to the standard
init+update+final mechanism.

levitte committed:
Fix EVP_Digest{Sign,Verify}Final() and EVP_Digest{Sign,Verify}() for provider only keys

For provider only keys where the initialization didn't catch, we may
end up crashing because the legacy code path didn't check that it had
support carefully enough.  This only happens if the caller didn't
check if initialization worked or not.

For the one-shot case, it's very simply handling the case where the
key has no legacy implementation an fall back to the standard
init+update+final mechanism.

levitte commented:
This fixes more test crashes for #10797

FdaSilvaYY commented on crypto/evp/m_sigver.c:
```diff
@@ -389,13 +394,18 @@ int EVP_DigestVerifyFinal(EVP_MD_CTX *ctx, const unsigned char *sig,
+        return 0;
+    }
+
+    if (pctx->pmeth->verifyctx)
```
nit: test against NULL ?

levitte replied:
Fixed

FdaSilvaYY commented on crypto/evp/m_sigver.c:
```diff
@@ -302,13 +302,18 @@ int EVP_DigestSignFinal(EVP_MD_CTX *ctx, unsigned char *sigret,
             r = pctx->pmeth->signctx(pctx, sigret, siglen, ctx);
         else {
-            EVP_PKEY_CTX *dctx = EVP_PKEY_CTX_dup(ctx->pctx);
+            EVP_PKEY_CTX *dctx = EVP_PKEY_CTX_dup(pctx);
```
nit: blank line after declaration.

levitte replied:
Done.

levitte committed:
squash! Fix EVP_Digest{Sign,Verify}Final() and EVP_Digest{Sign,Verify}() for provider only keys

While at it, EVP_DigestSignFinal() and EVP_DigestVerifyFinal() got a
slight code cleanup.

levitte commented:
Travis failure not relevant

levitte commented:
Final review, anyone?

levitte commented:
Merged.

8f020c3da571fd56c4a1ce16daad553193549a67 Fix EVP_Digest{Sign,Verify}Final() and EVP_Digest{Sign,Verify}() for provider only keys
